lib/net/rfb/frame_buffer.rb

```diff
@@ -0,0 +1,178 @@
+      @io = io
+      @w = w
+      @h = h
+      @bpp = bpp
+      @bypp = (bpp / 8.0).to_i
+      @wb = @w * @bypp
+      @data = "\x00" * @wb * @h
+
+      @cb_mutex = Monitor.new
+      @cb_cv = @cb_mutex.new_cond
+    end
+
+    # Set a way that server should use to represent pixel data
+    # @param [Hash] pixel format:
+    #  * {Net::RFB::PIX_FMT_BGR8}
+    #  * {Net::RFB::PIX_FMT_BGRA}
+    def set_pixel_format(format)
+      msg = [0, 0, 0, 0].pack('CC3')
+      begin
+        @io.syswrite msg
+
+        msg = [
+            format[:bpp],
+            format[:depth],
+            format[:bend],
+            format[:tcol],
+            format[:rmax],
+            format[:gmax],
+            format[:bmax],
+            format[:rshif],
+            format[:gshif],
+            format[:bshif],
+            0, 0, 0
+        ].pack('CCCCS>S>S>CCCC3')
+        return @io.syswrite msg
+
+      rescue
+        return nil
+      end
+    end
+
+    # Set way of encoding video frames.
+    # @param encodings [Array<Integer>] encoding of video data used to transfer.
+    #  * {ENC_RAW}
+    #  * {ENC_HEXTILE}
+    #  * {ENC_ZRLE}
+    def set_encodings(encodings)
+      num = encodings.size
+      msg = [2, 0, num].pack('CCS>')
+      begin
+        @io.syswrite msg
+        encodings.each do |e|
+          @io.syswrite([e].pack('l>'))
+        end
+      rescue
+        return nil
+      end
+    end
+
+    # Send request for update framebuffer.
+    # @param [Integer] inc incremental, request just difference
+    #  between previous and current framebuffer state.
+    # @param x [Integer]
+    # @param y [Integer]
+    # @param w [Integer]
+    # @param h [Integer]
+    def request_update_fb(inc, x: nil, y: nil, w: nil, h: nil)
+      ret = nil
+      @cb_mutex.synchronize do
+        if block_given?
+          @call_back = Proc.new { |data| yield data }
+        end
+
+        @inc = inc > 0
+        msg = [3, inc, x||0, y||0, w||@w, h||@h].pack('CCS>S>S>S>')
+        ret = @io.write msg
+
+        if block_given?
+          @cb_cv.wait
+          ret = @cb_ret
+          @cb_ret = nil
+        end
+      end
+      ret
+    rescue
+      return nil
+    end
+
+    # Handle VNC server response. Call it right after +fb_update_request+.
+    # @return [Array] type, (either framebuffer, "bell", +handle_server_cuttext+ or +handle_colormap_update+ results)
+    def handle_response(t)
+      case t
+      when 0 then
+        handle_fb_updated
+        if @call_back
+          @cb_mutex.synchronize do
+            cb = @call_back
+            @call_back = nil
+            begin
+              @cb_ret = cb.call @data
+            ensure
+              @cb_cv.broadcast
+            end
+          end
+        end
+        return [t, @data]
+      when 1 then
+        return [t, handle_colormap_update]
+      when 2 then
+        return [t, 'bell']
+      when 3 then
+        return [t, handle_server_cuttext]
+      else
+        return [-1, nil]
```

Comment:
It doesn't raise on errors, which is fine, so we could listen for a `[-1, nil]` response from the framebuffer, or we could let it raise/warn us of errors.